src/ring_theory/perfection.lean

```diff
@@ -0,0 +1,407 @@
+/-
+Copyright (c) 2020 Kenny Lau. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenny Lau
+-/
+
+import algebra.char_p
+import algebra.ring.pi
+import analysis.special_functions.pow
+import field_theory.perfect_closure
+import ring_theory.localization
+import ring_theory.subring
+import ring_theory.valuation.integers
+
+/-!
+# Ring Perfection and Tilt
+
+In this file we define the perfection of a ring of characteristic p, and the tilt of a field
+given a valuation to `ℝ≥0`.
+
+## TODO
+
+Define the valuation on the tilt, and define a characteristic predicate for the tilt.
+
+-/
+
+universes u v w
+
+open_locale nnreal
+
+/-- The perfection of a monoid `M`, defined to be the projective limit of `M`
+using the `p`-th power maps `M → M` indexed by the natural numbers, implemented as
+`{ f : ℕ → M | ∀ n, f (n + 1) ^ p = f n }`. -/
+def monoid.perfection (M : Type u) [comm_monoid M] (p : ℕ) : submonoid (ℕ → M) :=
+{ carrier := { f | ∀ n, f (n + 1) ^ p = f n },
+  one_mem' := λ n, one_pow _,
+  mul_mem' := λ f g hf hg n, (mul_pow _ _ _).trans $ congr_arg2 _ (hf n) (hg n) }
+
+/-- The perfection of a semiring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def comm_semiring.perfection (R : Type u) [comm_semiring R]
+  (p : ℕ) [hp : fact p.prime] [char_p R p] :
+  subsemiring (ℕ → R) :=
+{ zero_mem' := λ n, zero_pow $ hp.pos,
+  add_mem' := λ f g hf hg n, (frobenius_add R p _ _).trans $ congr_arg2 _ (hf n) (hg n),
+  .. monoid.perfection R p }
+
+/-- The perfection of a ring `R` with characteristic `p`,
+defined to be the projective limit of `R` using the Frobenius maps `R → R`
+indexed by the natural numbers, implemented as `{ f : ℕ → R | ∀ n, f (n + 1) ^ p = f n }`. -/
+def perfection (R : Type u) [comm_ring R] (p : ℕ) [hp : fact p.prime] [char_p R p] :
```

Comment:
Maybe this should even be called `subring.perfection`. Because we will also want `submodule.perfection` etc...
I'm not sure whether one of them deserves to be in the root namespace and the others don't. I would think that `subring.perfection` and `submodule.perfection` are on an "equal" footing in maths usage.

Reply:
can we get rid of `sub` and just call it `ring.perfection`?

Reply:
But then you can't use dot.notation

Reply:
I don't understand. You can only use dot notation if it has a semiring as **input** right?

Reply:
Ooh, you are completely right. I wasn't paying attention. So yes, let's shorten the namespace.